Source/displayinfo/DisplayInfo.cpp

```diff
@@ -0,0 +1,534 @@
+
+            if (index != _callbacks.end()) {
+                Trace("Unregistering displayinfo_updated_cb=%p", callback);
+
+                _callbacks.erase(index);
+            }
+        }
+
+        const string& Name() const { return _name; }
+        const bool IsValid() const { return _displayConnection != nullptr; }
+
+        bool IsAudioPassthrough() const
+        {
+            ASSERT(_displayConnection != nullptr);
+            return _displayConnection->IsAudioPassthrough();
+        }
+
+        bool Connected() const
+        {
+            ASSERT(_displayConnection != nullptr);
+            return _displayConnection->Connected();
+        }
+
+        uint32_t Width() const
+        {
+            ASSERT(_displayConnection != nullptr);
+            return _displayConnection->Width();
+        }
+
+        uint32_t Height() const
+        {
+            ASSERT(_displayConnection != nullptr);
+            return _displayConnection->Height();
+        }
+
+        Exchange::IConnectionProperties::HDRType HDR() const
+        {
+            ASSERT(_displayConnection != nullptr);
+            return _displayConnection->Type();
+            ;
+        }
+
+        Exchange::IConnectionProperties::HDCPProtectionType HDCPProtection() const
+        {
+            ASSERT(_displayConnection != nullptr);
+            return _displayConnection->HDCPProtection();
+        }
+
+    private:
+        DisplayInfoAdministrator& _parent;
+        mutable uint32_t _refCount;
+        const string _name;
+        Exchange::IConnectionProperties* _displayConnection;
+        Core::Sink<Notification> _notification;
+        Callbacks _callbacks;
+    };
+
+public:
+    ~DisplayInfoAdministrator()
+    {
+        ASSERT(_clients.size() == 0);
+
+        for (auto client : _clients) {
+            Trace("Closing %s", client->Name().c_str());
+            delete client;
+        }
+
+        if (_comChannel->IsOpen()) {
+            _comChannel->Close(RPC::CommunicationTimeOut);
+            Trace("Closed COM channel %d", _comChannel->ConnectionId());
+        }
+    }
+
+    DisplayInfoAdministrator(const DisplayInfoAdministrator&) = delete;
+    DisplayInfoAdministrator& operator=(const DisplayInfoAdministrator&) = delete;
+
+    void Register(DisplayInfo* client)
+    {
+        std::list<DisplayInfo*>::iterator index(std::find(_clients.begin(), _clients.end(), client));
+
+        if (index == _clients.end()) {
+            if (_clients.size() == 0) {
+                _comChannel->Open(RPC::CommunicationTimeOut);
+                Trace("Opened COM channel %d", _comChannel->ConnectionId());
+            }
+
+            client->Initialize(_comChannel);
+
+            if (client->IsValid() == true) {
+                Trace("Added client %p to registery[%d]", client, _clients.size());
+                _clients.push_back(client);
+            }
+
+            if (_clients.size() == 0) {
```

Comment:
If we add it to the previous if there is a issue when a user requests a non-existing implementation after requesting a valid one, it will also close the connection of the first one.